Source/Test/Rewriting/Passes/AssemblyRewriter.cs

```diff
@@ -468,124 +516,36 @@ protected static TypeDefinition Resolve(TypeReference type)
-        /// <param name="genericType">The generic type to instantiate.</param>
-        /// <param name="genericArgs">The generic arguments needed to instantiate the generic type.</param>
-        /// <returns>The new generic instance type.</returns>
-        protected static GenericInstanceType ImportGenericTypeInstance(ModuleDefinition module, TypeReference genericType,
-            params TypeReference[] genericArgs)
+        public static TypeReference MakeGenericType(TypeReference self, params TypeReference[] arguments)
         {
-            TypeReference typeDef = genericType.Resolve();
-            if (!typeDef.HasGenericParameters)
+            if (self.GenericParameters.Count != arguments.Length)
             {
-                throw new InvalidOperationException(string.Format("Type {0} is not generic", genericType));
+                throw new ArgumentException();
             }
 
-            typeDef = module.ImportReference(typeDef);
-            var instance = new GenericInstanceType(typeDef);
-            for (int i = 0; i < typeDef.GenericParameters.Count; i++)
+            var instance = new GenericInstanceType(self);
+            foreach (var argument in arguments)
             {
-                var p = typeDef.GenericParameters[i];
-                if (p.Position < genericArgs.Length)
-                {
-                    GenericParameter parameter = new GenericParameter(p.Name, typeDef);
-                    instance.GenericParameters.Add(parameter);
-                    instance.GenericArguments.Add(genericArgs[p.Position]);
-                }
-                else
-                {
-                    throw new InvalidOperationException(string.Format("Not enough generic arguments to instantiate type {0}", genericType));
-                }
+                instance.GenericArguments.Add(argument);
             }
 
             return instance;
         }
 
-        /// <summary>
-        /// Create a GenericInstanceMethod from the given generic method and the given generic arguments.
-        /// </summary>
-        /// <param name="module">The module we are operating on.</param>
-        /// <param name="genericMethod">A generic method to instantiate.</param>
-        /// <param name="genericArgs">The combined generic arguments for the declaring type (if it is generic) and for the method.</param>
-        /// <returns>The new method reference.</returns>
-        /// <remarks>
-        /// This method can also handle the case where the DeclaringType is also generic.
-        /// Simply pass the combined generic args for the declaring type and the method.
-        /// </remarks>
-        protected static MethodReference ImportGenericMethodInstance(ModuleDefinition module, MethodReference genericMethod,
-            params TypeReference[] genericArgs)
+        public static MethodReference MakeGenericMethod(MethodReference self, params TypeReference[] arguments)
```

Comment:
Wow man, these two methods (`MakeGenericMethod` and `MakeGenericType`) look so much simpler now!